dracut/modules.d/99kiwi-dump-reboot/kiwi-dump-reboot-system.sh

```diff
@@ -33,17 +33,40 @@ function boot_installed_system {
     fi
 }
 
+function mount_ramdisk_sysroot {
+    local image_target
+    local root_uuid
+    local boot_options
+
+    boot_options="/config.bootoptions"
+
+    if [ ! -e "${boot_options}" ]; then
+        die "Missing ${boot_options} file"
+    fi
+
+    root_uuid=$(
+        while read -r -d ' ' opt; do echo "${opt}";done < "${boot_options}" |\
+        grep root= | cut -f2- -d=
+    )
+
+    if [ -z "${root_uuid}" ]; then
+        die "'root=' argument not found in ${boot_options}"
+    fi
+
+    image_target=$(get_selected_disk)
+    kpartx -s -a "${image_target}"
+    systemd-mount --fsck=no --no-block "${root_uuid}" /sysroot
```

Comment:
Well `systemd-mount` seems still not appropriate to me:

- the transient mount unit created here has no dependencies on `initrd-root-fs.target`, so you can't really tell for sure when PID1 will schedule it.
- since `--no-block` is used you actually don't even wait for the transient unit to be added to the list of current jobs before returning from `mount_ramdisk_sysroot()`.

I don't really understand why you want to use `systemd-mount` here especially since this tool is supposed to be used when file-systems needs to be mounted asynchronously. 

Why don't you simply use mount(8) here ?

PS: I would also add a comment explaining how /sysroot is mounted as it's not really obvious ;)

Reply:
> Why don't you simply use mount(8) here ?

I guess this was done to let systemd control the mount process for root. From my perspective the asynchronous behavior is ok at that point. The problem with the former generator for the sysroot unit was that the generator was depending on a file to be downloaded from the network in case of the remote (PXE) use case. If the generator was called before the code here has downloaded the file from the network no sysroot unit was created. With the changes David did we make sure the systemd-mount process is called not before any needed data is present. After that it does not play a role when PID1 schedule it.

I think it is part of the design that systemd should know about /sysroot. We can of course just mount it but wouldn't it be better if systemd controls this ?

I agree regarding the --no-block. I think that should not be used since I assume the call to return as soon as systemd has created the unit file or am I mistaken ?

Thanks much for your feedback

Reply:
It was a stupid idea from my side to read a file within system-generators/dracut-kiwi-ramdisk-generator code when I could not be sure this file exists no matter at what time the generator code is called.

Reply:
@fbuihuu thanks for sharing you view, this is really valuable.

> I don't really understand why you want to use systemd-mount here especially since this tool is supposed to be used when file-systems needs to be mounted asynchronously.

The motivation of using `systemd-mount` as @scheafi said was to make systemd aware of `/sysroot` mount, as it thought this is something desirable. Probably for a ramdisk deployment this not that much of an issue.

> the transient mount unit created here has no dependencies on initrd-root-fs.target, so you can't really tell for sure when PID1 will schedule it.

That's correct but there is the `dracut-mount.service` right after `dracut-pre-mount.service` that actually waits for /sysroot being mounted, otherwise it does not continue as far as I could understand by reading the code.

> since --no-block is used you actually don't even wait for the transient unit to be added to the list of current jobs before returning from mount_ramdisk_sysroot()

I added it to sort out a deadlock, apparently the transient mount unit gets scheduled after the `pre-mount-dracut.service`, so this was causing a dead lock.

So I see four options to proceed:

1. Leave it as is in this PR
2. Run simple mount command in `mount_ramdisk_sysroot()` method
3. Include a mount hook (they are executed by `dracut-mount.service`) that only operates on `rd.kiwi.ramdisk` argument presence
4. In `mount_ramdisk_sysroot()` we create our own custom `sysroot.mount` unit manually (with the dependencies that suit our needs to not mess any running transaction) and run `systemctl daemon-reload`.

Probably the second and the third are the simplest ones and for a ramdisk deployment this might be sufficient.

Reply:
I would go with option 2 for this use case. I don't see how a dracut mount hook would be called when there is no /sysroot mountpoint. Isn't that a chicken&egg problem.

Regarding mount points: These are usually all handled by systemd through unit files. There is an fstab generator, there is a sysroot generator and so on. systemd does this to have control over all mount points of the system, which is a good concept. Running a manual mount circumvents this concept but is not a problem when the system was deployed into a ramdisk. That's because no data is persistent and on reboot everything is gone and you need to start from scratch. So it's ok to manually mount /sysroot in this ramdisk case.

But to be honest it's not nice and it's an exception and it would be great if we wouldn't need that.

If the deployment of an image happens into a real disk, not a ramdisk, the standard sysroot target created by systemd through the information of root=... works without any problems. The chain of events in this case is like this:

* dump of raw disk image to a storage device
* kexec into the new system
* udev automatically recognizes the new devices, e.g root partition
* udev creates all devices according to its rules
* the systemd generator for sysroot identifies the device provided via the root= kernel cmdline and is able to create a mount unit

done

Now the process using a ramdisk device

* dump of raw disk image to a brd device. kiwi has to manually load the brd driver prior dump and does this if rd.kiwi.ramdisk is set. Once brd module is loaded some /dev/ramX devices exists
* kexec boot is not possible because that would wipe the ramdisk contents. Therefore the boot into the system must be done from the environment that has dumped the image. In order to get partition devices kiwi calls: ```kpartx -s -a "${image_target}"```. This call kicks udev and all devices e.g root partition are created
* For some reason the systemd generator for sysroot has not created a unit file. This was the reason why I added kiwi-ramdisk-deployment-generator in pre-mount formerly

Maybe we should look for a different solution to the problem. Is it possible to tell systemd to re-run the generator for sysroot after we called ``kpartx -s -a "${image_target}"``` ? I was thinking at this code path:

```bash

... make sure the config file exists

if getargbool 0 rd.kiwi.ramdisk; then
    # For ramdisk deployment a kexec boot is not possible...
    image_target=$(get_selected_disk)
    kpartx -s -a "${image_target}"
    # make sure systemd sysroot generator is kicked...
else
    ...
fi
```

and no extra mount code from our side needed

Thoughts ?

Reply:
> I would go with option 2 for this use case. I don't see how a dracut mount hook would be called when there is no /sysroot mountpoint. Isn't that a chicken&egg problem.

This does not relay on mount units, this executes custom user provided hook (as the pre-mount hook) scripts. So, if we actually mount that manually, which I am not a fan of, I believe this is the point where it should be mounted. See dracut code here 
https://github.com/dracutdevs/dracut/blob/7c2521091f30d7c14caaf96850208afa2fe1d811/modules.d/98dracut-systemd/dracut-mount.sh#L16-L29

dracut-mount.service executes all mount hooks one after the other until sysroot is mounted or until the timeout is reached. So I assume if we follow the path of manually mounting sysroot, dracut would expect us to do so in a mount hook. Anyway I am opened to other solutions based on systemd as I am not a fan of this solution.

>  systemd does this to have control over all mount points of the system, which is a good concept

yes I agree, that's why I initially tried the `systemd-mount` approach.

> For some reason the systemd generator for sysroot has not created a unit file. This was the reason why I added kiwi-ramdisk-deployment-generator in pre-mount formerly

It is not created because there is no `root=` argument to pass. As a result systemd continues without having any dependency on `sysroot.mount` and it is `dracut-mount.sh` and probably other steps that check if `/sysroot` is mounted or not, without relaying in systemd status.

> Maybe we should look for a different solution to the problem. Is it possible to tell systemd to re-run the generator for sysroot after we called ``kpartx -s -a "${image_target}"``` ? I was thinking at this code path:

This is the current released approach. The generator was currently re executed in pre-mount hook and modifying the initrd-root-fs.target transaction (which was about to finalize or even finalized at that point). I'd say that if we want to go with the systemd way we should do something like creating a `sysroot.mount` unit (not from a generator, just create it within the pre-mount hook itself) without the common `intird-fs-root.target` dependecy and make it start before `dracut-mount.service`, so we are certain that at least during the execution of `dracut-mount.service` the mount unit will be handled.

Between, I found another potential issue on the current generator for sysroot on ramdisk deployments. The unit file does not include `DefaultDependencies=no` otherwise dependencies are included on earlier services [see the docs](https://www.freedesktop.org/software/systemd/man/systemd.mount.html#Default%20Dependencies).

So I agree on using a sysroot.mount, but I don't think we should do that based on a generator. I would just directly dump the `sysroot.mount` file within the pre-mount hook and make sure the dependencies of this unit are sane in that boot stage to no mess with ongoing transactions.

Reply:
>So I agree on using some sysroot.mount created by us, but I am not sure doing that using a generator is the best way to do it. I would just directly dump the sysroot.mount file within the pre-mount hook and make sure the dependencies of this unit are sane in that boot stage to no mess with ongoing transactions. Basically this is my fourth option.

For me this would make sense as it is the closest approach to the default behavior